### PR TITLE
Enh/add auth

### PR DESCRIPTION
**Pull request by MBueschelberger**
# Description:

This PR shall represent a draft for adding a potential authentication layer based on [`fastapi.security`](https://fastapi.tiangolo.com/tutorial/security/) to the FastApi-app.

For this procedure, the `OTEAPI_AUTH_DEPS`-env can be enabled, which underlies a similar syntax like the `OTEAPI_PLUGIN_PACKAGES`-variable. The major difference is that the single modules are referencing to the individual class of the Python-package-regex: `my_package.my_module.my_class`.

*Of course*, one has to be careful hiding the `OTEAPI_AUTH_DEPS`-variable to the outside world, since it may lead to similar security concerns such as the `OTEAPI_PLUGIN_PACKAGES`-env.

An example of the overall variable my look like that:
`my_module.my_class | my_other_module.my_other_class`

When this example variable will be parsed during start-up of the app, `my_class` and `my_other_class` will be imported and hence result in:

`FastApi(dependencies=[Depends(my_class()), Depends(my_other_class))]`

This gives the flexibility that a developer can come up with its own authentication scheme. 

Moreover, `my_class` or `my_class` might be a childclass of the common classes from the `fastapi.security`-module such as `HTTPBasic` or  `HTTPBearer` and can provide a callback-function by overriding the `__call__`-method:

``` 
MyHttpAuth(HTTPBearer):
        async def __call__(self, request: Request) -> str:
                auth = await super().__call__(request=request)
                my_callback_function(auth.credentials)
                return auth.credentials
```

The token from the auth.credentials may then be further used for authentication to `OntoRec` etc.

I am keen towards your feedback and suggestions for improvement!

## Type of change:

- [ ] Bug fix.
- [x] New feature.
- [ ] Documentation update.

## Checklist for the reviewer:

This checklist should be used as a help for the reviewer.

- [ ] Is the change limited to one issue?
- [ ] Does this PR close the issue?
- [x] Is the code easy to read and understand?
- [ ] Do all new feature have an accompanying new test?
- [ ] Has the documentation been updated as necessary?

### CODE REVIEW

File: app/main.py [hunk 1]
```diff
@@ -85,6 +93,30 @@ def create_app() -> FastAPI:
...
+    "function in order to fetch dependencies for authentication"
+    deps = os.environ.get("OTEAPI_AUTH_DEPS")
+    if deps:
+        mods = [
+            module.replace(" ", str()).rpartition(".") for module in deps.split("|")
+        ]
+        try:
+            imports = [
+                getattr(import_module(module), classname)
+                for (module, dot, classname) in mods
```
Comment (CasperWA) [inline app/main.py:106]: I think, traditionally, one writes out the module as you suggest (`package_name.module.sub_module`), but class names are added after a colon, e.g., `package_name.module.sub_module:MyClass`.
So I'd probably go this route instead?

Also, using `rpartition` is nice, but dangerous. What if the class is imported from the root module/package name? I.e., `package_name:MyClass`? Then your import statement would be `getattr(import_module(""), "package_name")` in the current version. I think with the colon this should all work out, as you can `rpartition` with the colon and all this should be fine :)

Also:
```suggestion
                getattr(import_module(module), classname)
                for (module, _, classname) in modules
```

File: app/main.py [hunk 2]
```diff
@@ -85,6 +93,30 @@ def create_app() -> FastAPI:
...
+            ]
+        except Exception as error:
+            raise error
```
Comment (CasperWA) [inline app/main.py:109]: This is the same as just not using a try/except - but I guess your intention here is to figure out what specific exceptions to catch and raise another specific exception with a custom message?

File: docker-compose_dev.yml [hunk 3]
```diff
@@ -14,6 +14,7 @@ services:
       OTEAPI_prefix: "${OTEAPI_prefix:-/api/v1}"
       PATH_TO_OTEAPI_CORE:
       OTEAPI_PLUGIN_PACKAGES:
+      OTEAPI_AUTH_DEPS:
```
Comment (CasperWA) [inline docker-compose_dev.yml:17]:
```suggestion
      OTEAPI_AUTH_DEPENDENCIES:
```

File: app/main.py [hunk 4]
```diff
@@ -24,6 +27,10 @@
     from typing import Any, Dict
 
 
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.INFO)
+
+
 class AppSettings(RedisSettings):
     """Redis settings."""
```
Comment (CasperWA) [inline app/main.py:36]:
```suggestion

    AUTH_DEPENDENCIES: str = Field("", description="List of FastAPI dependencies for authentication features.")
```
You can even consider adding the parsing of this variable into this class as a `pydantic.validator` class method. But this is fine for now, for sure :)

**Review by CasperWA — CHANGES_REQUESTED** [review]
Some suggested changes here, but I really like this idea of passing in the authentication classes/functions in this way.

File: app/main.py (hunk 1 above)
Comment (MBueschelberger) [reply app/main.py:113]: Very good hint. I will apply your suggestion with the colon

File: app/main.py [hunk 5]
```diff
@@ -85,6 +93,30 @@ def create_app() -> FastAPI:
...
+            imports = [
+                getattr(import_module(module), classname)
+                for (module, dot, classname) in mods
+            ]
+        except Exception as error:
+            raise error
+        logger.info(
+            "Imported the following dependencies for authentication: %s", imports
+        )
+        dependencies = [Depends(dependency()) for dependency in imports]
```
[the remark this answers is not in the document]
Comment (MBueschelberger) [reply app/main.py:120]: I am not sure here what is the convention. For testing, I used a derivation of the [HTTPBearer](https://gitlab.cc-asp.fraunhofer.de/MarketPlace/platform/modules/accesscontrolpackage/-/blob/enh/fastapi/src/accesscontrol/frameworks/fastapi.py#L19) from the MarketPlace which overriding the `__call__`-method for the callback to the OpenID-Client. Do you think it is more convenient to perform the callback in the `__init__`-method or even the `__post__init__` so that the dependency does not have to called in `Depends`?

**Comment by MBueschelberger** [comment]
@CasperWA :

I added a [unit test for checking](https://github.com/MBueschelberger/oteapi-services/blob/c00e8be2343b63d3c3cdcfce98c1a93420e875a4/tests/routers/test_transformation.py#L43) the retrieval of the [token from the configs or the request headers](https://github.com/MBueschelberger/oteapi-services/blob/c00e8be2343b63d3c3cdcfce98c1a93420e875a4/app/routers/transformation.py#L45). While running the pytests in `docker_compose_dev.yml` manually via cli in the container, all tests exit successfully. However, the [`SecretStr` seems to be censored to "******"](https://github.com/EMMC-ASBL/oteapi-services/actions/runs/3694795423/jobs/6256406514#step:3:944) only during the docker-build. Is it maybe related that the `DummyCache` for the pytests are using different [`json_encoders` ](https://pydantic-docs.helpmanual.io/usage/types/#secret-types) for `SecretStr` in different environments?

FYI: I made the [exposure of the redis-cache endpoint optional](https://github.com/MBueschelberger/oteapi-services/blob/c00e8be2343b63d3c3cdcfce98c1a93420e875a4/app/main.py#L94), since I think this might not be desired in various prodcution scenarios as explained in the `AppSettings` [here](https://github.com/MBueschelberger/oteapi-services/blob/c00e8be2343b63d3c3cdcfce98c1a93420e875a4/app/main.py#L36).

**Comment by MBueschelberger** [comment]
@CasperWA, I think all threads are resolved. As soon as [this](https://github.com/EMMC-ASBL/oteapi-core/pull/222) PR is merged, we can update [this](https://github.com/MBueschelberger/oteapi-services/blob/ed1468b9ceae248f09c611c837764816df9b6408/requirements.txt#L34) line in the `requirements.txt` and the branch would be ready to be merged as well here.

**Comment by CasperWA** [comment]
> @CasperWA, I think all threads are resolved. As soon as [this](https://github.com/EMMC-ASBL/oteapi-core/pull/222) PR is merged, we can update [this](https://github.com/MBueschelberger/oteapi-services/blob/ed1468b9ceae248f09c611c837764816df9b6408/requirements.txt#L34) line in the `requirements.txt` and the branch would be ready to be merged as well here.

It's a reality! You can update to minimum v0.3.0, i.e.: `~=0.3.0`.

**Comment by CasperWA** [comment]
Hmm, I can't immediately penetrate the test error messages. Do you mind if I take a stab at it locally and commit a fix, if I find one? Otherwise I'll just let you know the issue and you could commit :)

**Comment by MBueschelberger** [comment]
Hi @CasperWA,

I can reproduce locally. But I am puzzled what is the difference to the previous dev-version of the  `SecretConfig` and the current one from `oteapi-core==0.0.3`.  I will try yo investigate

**Comment by MBueschelberger** [comment]
@CasperWA, I have figured out the issue.

The routes in the `redisadmin` was doing a type check while returning the data from the `DummyCache`. In my example I use the `redis`-endpoint in order to test whether the `token` has been retrieved through the config or through the request-headers.

Nevertheless, the `DummyCache` should also return a `str`, because the sessions/configs are serialized via the json-method in the routers of each single strategy (see `json.dumps` in the `test_data` for the `pytest`). This leads us to the opportunity, that we can neglect the `json.dumps` in the `get`-method of the `DummyCache` and the conversion from JSON-`str` to a `dict` works properly via `json.loads` in all routers.

File: tests/conftest.py [hunk 6]
```diff
@@ -26,9 +26,7 @@ async def set(self, id, data) -> None:
 
     async def get(self, id) -> dict:
         """Mock `get()` method."""
-        import json
-
-        return json.dumps(self.obj[id])
+        return self.obj[id]
```
Comment (CasperWA) [inline tests/conftest.py:29]: Maybe it should be `json.loads(json.dumps(self.obj[id]))`? Just to ensure it can be JSON encoded as it would be in the actual endpoint?

File: tests/conftest.py [hunk 7]
```diff
@@ -54,35 +52,45 @@ def top_dir() -> Path:
...
+                "mappingType": "mapping/demo",
+                "prefixes": {":": "<http://namespace.example.com/ns#"},
+                "triples": [[":a", ":has", ":b"]],
+                "configuration": {},
+            }
+        ),
         # sessions
-        "1": {"foo": "bar"},
-        "2": {"foo": "bar"},
+        "1": json.dumps({"foo": "bar"}),
```
Comment (CasperWA) [inline tests/conftest.py:84]: Ah, you moved the `json.dumps()` into here instead. Fair.

Maybe it'd be nicer to store this big dict in a temporary variable then instead, and then return a new dict where we run through that variable passing the values through `json.dumps()`. It makes it slightly easier to add new testdata... Although, I suppose it's the same.
Up to you then :)

Comment (MBueschelberger) [reply tests/conftest.py]: Ah, I see. You already provided a solution for this. I initially thought about a `pytest.parametrize`-decorator in order to pass the DummyCache-data in the test_**.py itself. But it might also be a bit of overkill and messyin the end ...

File: docker-compose.yml [hunk 8]
```diff
@@ -10,7 +10,10 @@ services:
       OTEAPI_REDIS_HOST: redis
       OTEAPI_REDIS_PORT: 6379
       OTEAPI_prefix: "${OTEAPI_prefix:-/api/v1}"
+      OTEAPI_INCLUDE_REDISADMIN: "${OTEAPI_INCLUDE_REDISADMIN:-False}"
+      OTEAPI_EXPOSE_SECRETS: "${OTEAPI_EXPOSE_SECRETS:-True}"
```
Comment (CasperWA) [inline docker-compose.yml:14]: Wait. So by default you don't want to include the Redis admin endpoint (for security reasons) but _do_ want to expose secrets? :/

**Review by CasperWA — COMMENTED** [review]
Should there be a similar test for dataresource as well?

Also, I cleaned up the `conftest` file according to my comments above. I hope this is fine?

**Comment by CasperWA** [comment]
@MBueschelberger Could you comment on the last part, Then this can be merged and I can look at the OTELib PR that depends on this.

File: docker-compose.yml (hunk 8 above)
Comment (MBueschelberger) [reply docker-compose.yml:14]: Well, the redis admin is not exposed *because* the secrets are exposed.

If the secret strings for the `oteapi.models` are not exposed, they cannot be read in the strategies anymore, since they are censored during serialization. This pretty much makes them useless if you want to use them for connecting to an external host. 

If they are exposed and the redis admin is enabled, you may simply reveal all user's in a configuration through the redis endpoint. 

For this reason, the chance, that you may remotely inspect the redis-cache from other's users sessions, is lowered since you only can access the redis-cache from the strategies, and not directly through the services anymore.

In other words, the motivation is that when you put sensitive/general information into the cache, you cannot reveal it by chance if you *only* got the config-id, but it still can be used by the strategies internally.

Since I did not find any straight-forward user management for redis through an oauth-scheme, I think this is one of the only ways how to deal with this scenario for the moment.

Does this make sense to you?

Comment (CasperWA) [reply docker-compose.yml:14]: This makes sense. Essentially what we need is to hash/encrypt the secrets before storing in redis. That should solve it. The strategies can then get some interface to decrypting the secrets they may need. Or it's done from the service before invoking the strategies. Not _very_ important, but it should circumvent this issue.

However that's something for another day and another PR.

**Review by CasperWA — APPROVED** [review]

**Comment by CasperWA** [comment]
@MBueschelberger merge at will :)
